src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,153 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    static class Node<T> {
+        private T val;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(T val, Node<T> prev) {
+            this.val = val;
+            this.prev = prev;
+        }
+
+        public Node(T value, Node<T> prev, Node<T> next) {
+            this.prev = prev;
+            this.val = value;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> node = new Node<T>(value, tail);
+        if (size == 0) {
+            head = node;
+        } else {
+            tail.next = node;
+        }
+        tail = node;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("The index " + index + " is invalid");
+        }
+        if (findNode(index) == null) {
+            add(value);
+        } else {
+            Node<T> prev = findNode(index).prev;
+            Node<T> newNode = new Node<>(value, prev, findNode(index));
+            if (index == 0) {
+                head = newNode;
+            } else {
+                prev.next = newNode;
+                findNode(index).prev = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexValidity(index);
+        return findNode(index).val;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexValidity(index);
+        Node<T> node = findNode(index);
+        T oldValue = node.val;
+        node.val = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexValidity(index);
+        Node<T> node = findNode(index);
+        unlink(index, node);
+        return node.val;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> node = head;
+        int index = -1;
+        for (int i = 0; i < size; i++) {
+            if (object == node.val || object != null && object.equals(node.val)) {
+                index = i;
+                break;
+            } else {
+                node = node.next;
+            }
+        }
+        if (index == -1) {
+            return false;
+        }
+        unlink(index, node);
```

Comment:
I recommend you rewrite this method like this:
```suggestion
        for (Node<T> node = head; node != null; node = node.next) {
             if (node.item == object || object != null && object.equals(node.item)) {
                 unlink(node);
                 return true;
             }
         }
         return false;
```